### PR TITLE
Add missing Model Builder API

### LINKED ISSUE #33287: Add IConventionEntityTypeBuilder API for UseSqlReturningClause

### LINKED ISSUE #33739: OwnedNavigationBuilder.HasIndex lacks overload with name argument
Hello everyone!

I have a need to remap a bunch of fields into an owned entity. Some of the fields have multiple indexes on them.
With EntityTypeBuilder it works like a charm after implementation of https://github.com/dotnet/efcore/issues/21089.

But I found out that OwnedNavigationBuilder lacks HasIndex method overload with name argument, tho they're using the same InternalEntityTypeBuilder under the hood.
This leads to the same problem that latest index overrides every index before it.

I looked through PR: https://github.com/dotnet/efcore/pull/21115 and didn't find any mentions of owned entities.

I wonder if there's any reason it was not implemented or just forgotten about?
If it duplicates already existing question, please point to it, I wasn't able to find anything related on my own

EF Core version: 8.0.2
Database provider: Npgsql 8.0.2
Target framework: .NET 8.0

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Add OwnedNavigationBuilder.HasIndex with name
Add IConventionEntityTypeBuilder API for UseSqlReturningClause and UseSqlOutputClause
Rename some `CanSet*` methods to `CanHave*`

Fixes #33739
Fixes #33287

### CODE REVIEW

**Review by sonika1311 — APPROVED** [review]

File: src/EFCore.Cosmos/Extensions/CosmosEntityTypeBuilderExtensions.cs [hunk 1]
```diff
@@ -632,7 +632,7 @@ public static EntityTypeBuilder<TEntity> UseETagConcurrency<TEntity>(this Entity
     public static IConventionEntityTypeBuilder? IncludeRootDiscriminatorInJsonId(
         this IConventionEntityTypeBuilder entityTypeBuilder, bool? includeDiscriminator, bool fromDataAnnotation = false)
     {
-        if (!entityTypeBuilder.CanSetIncludeRootDiscriminatorInJsonId(includeDiscriminator, fromDataAnnotation))
+        if (!entityTypeBuilder.CanIncludeRootDiscriminatorInJsonId(includeDiscriminator, fromDataAnnotation))
```
Comment (roji) [inline src/EFCore.Cosmos/Extensions/CosmosEntityTypeBuilderExtensions.cs:635]: I think I missed the design discssion on these - so are we departing from the CanSet* naming convention? Don't these methods determine whether the option can be set, as opposed to whether e.g. the root discriminator can actually be included in the JSON ID (it always can be, no?).

(in any case am approving, if I've just missed/forgot the team decision ignore me)

File: src/EFCore/Metadata/Builders/OwnedNavigationBuilder.cs [hunk 2]
```diff
@@ -334,6 +334,23 @@ public virtual IndexBuilder HasIndex(params string[] propertyNames)
...
 
+    /// <summary>
+    ///     Configures an index on the specified properties and with the given name.
+    ///     If there is an existing index on the given list of properties and with
+    ///     the given name, then the existing index will be returned for configuration.
+    /// </summary>
+    /// <param name="propertyNames">The names of the properties that make up the index.</param>
+    /// <param name="name">The name to assign to the index.</param>
+    /// <returns>An object that can be used to configure the index.</returns>
+    public virtual IndexBuilder HasIndex(
```
Comment (roji) [inline src/EFCore/Metadata/Builders/OwnedNavigationBuilder.cs:345]: nit: put on a single line - I think we wrap only if we reach the line char limit?

**Review by roji — APPROVED** [review]

File: src/EFCore.Cosmos/Extensions/CosmosEntityTypeBuilderExtensions.cs (hunk 1 above)
Comment (AndriySvyryd) [reply src/EFCore.Cosmos/Extensions/CosmosEntityTypeBuilderExtensions.cs:635]: We haven't. This is just something that Arthur mentioned in the API review, I'll revert

File: src/EFCore/Metadata/Builders/OwnedNavigationBuilder.cs (hunk 2 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/Builders/OwnedNavigationBuilder.cs:345]: I'd need to change all methods in this file, better done in code cleanup

**Pull request merged**